[export] min/max ranges for dim hints

pianpwk opened the pull request:
Differential Revision: D71522032

Adds min/max ranges to Dim.AUTO/DYNAMIC/STATIC, so users can do `Dim.AUTO(min=2, max=2048)`.

pianpwk commented on torch/export/dynamic_shapes.py:
```diff
@@ -69,6 +72,14 @@ def DYNAMIC():
+        assert min is None or min >= 0, "min must be non-negative"
+        assert max is None or max >= 0, "max must be non-negative"
+        assert min is None or max is None or min <= max, "min must be <= max"
+        return _DimHint(self.type, min=min, max=max, _factory=False)
```
this is sneaky, but adding this `_factory` field so users can do `Dim.AUTO`, `Dim.AUTO()`, or `Dim.AUTO(min=2, max=2048)` and get the same object class, but raise an error if they do `Dim.AUTO()()`

angelayi commented on torch/_export/non_strict_utils.py:
```diff
@@ -394,24 +399,53 @@ def make_constraints(
+                        trace_vr & user_vr
+                    else:
+                        range_constraints[d.node.expr] &= user_vr
+                except torch.utils._sympy.value_ranges.ValueRangeError as exc:
+                    raise ValueError(
+                        f"Received user-specified min/max range of [{dim.min}, {dim.max}], "
+                        f"which conflicted with the inferred min/max range of [{trace_vr.lower}, {trace_vr.upper}] "
+                        f"from model tracing, for inputs{pytree.keystr(flat_paths[input_index])}.shape[{i}]."
+                    ) from exc
```
I'm a little confused when this error happens. My understanding is that this function is just getting range constraints, but this error seems like a ConstraintViolationError which I would've expected to be thrown earlier?

pianpwk replied:
Yeah, this is a bit weird, but I couldn't find a good way to feed these ranges into produce_guards, so this won't show up for the constraint solver / suggested fixes. We could improve and print out all the violations we hit, and not just the first though.

I did want to keep this separate, since in constraint violations it's a lot of user work to track down where the min/max bound guards are emitted, and here we just know it's from the user-specification.

Is there any error message / UX you'd suggest though? very open

ColinPeppler commented on test/export/test_export.py:
```diff
@@ -2371,6 +2371,76 @@ def forward(self, x, y, z):
+                assert y.shape[0] <= 32
+                return x[6:], y + 2
+
+        x, y = torch.randn(8), torch.randn(8)
+
+        # conflict with lower bound
+        shapes = torch.export.ShapesCollection()
+        shapes[x] = [Dim.DYNAMIC(max=5)]
+        with self.assertRaisesRegex(
+            ValueError,
+            r"Received user-specified .* \[None, 5\], which conflicted with the inferred .*"
+            r"\[6, int_oo\] .*, for inputs\['xs'\]\['data'\]\[0\]\[0\]\.shape\[0\]",
```
questions from a UX perspective (it's okay if no good answer)
- why is this the error we get?
- would it be possible for the error to mention that max=5, but the x is `torch.Size([8])`?

pianpwk replied:
Here model tracing figured out the valid range is [6, inf], because of the `x[6:]` call, while user specified [0, 5]. These ranges don't overlap, so we raise this error.

We could definitely check the sample input shapes too, but that'd be a separate check (we'd check sample input conflicts first, then check inferred range vs. user range).

I'm curious (totally open question), is there any error message you'd like to see / UX you'd like to have here? I kind of wrote this generically.

ColinPeppler replied:
Thanks for the explanation! I see so I see it's smth like if `[user_specified_min, user_specified_max]` is not a subset of the `[inferred_min, inferred_max]`, then Export throws.

> We could definitely check the sample input shapes too, but that'd be a separate check (we'd check sample input conflicts first, then check inferred range vs. user range).

Yeah, that makes sense. I was expecting whether I'm using Dim Hints or using Named Dim, the type of checks and when the checks happen would be the same.

>  is there any error message you'd like to see / UX you'd like to have here?
I think the current error message is fine! Though, I'm wondering if it's more expedient to throw an error, at the start, mentioning the sample input shapes don't match with the user-specified range.

pianpwk commented on torch/_export/non_strict_utils.py:
```diff
@@ -359,17 +360,21 @@ def make_constraints(
-    range_constraints = {
-        symbol: inline_constraints[symbol] for symbol in inline_constraints
-    }
+    range_constraints = defaultdict(lambda: ValueRanges(0, int_oo)) | inline_constraints
```
I wanted this to be a defaultdict haha

angelayi replied:
more fancy python

tugsbayasgalan replied:
This is pretty cool lol

pianpwk commented on torch/_export/non_strict_utils.py:
```diff
@@ -394,24 +399,53 @@ def make_constraints(
-                dim = shape_spec[i] if shape_spec else None
                 if dim is None or isinstance(dim, _DimHint):
-                    range_constraints[d.node.expr] = shape_env.var_to_range[
+                    range_constraints[d.node.expr] &= shape_env.var_to_range[
```
when we go through the placeholder symbols & dynamic shapes specs, we might see the same symbols over and over, with new min/max ranges, e.g.
```
sizes: [s0, s1], [s0, 4]
shapes: [AUTO(min=4), AUTO], [AUTO(max=32), None]
```

each time we see a symbol (s0), we wanna refine it with both the inferred range (from ShapeEnv), and the user-specified range (dim.min/max). The `&` just takes the intersection of ranges.

angelayi replied:
fancy

angelayi replied:
is it ever possible that you get one that says AUTO(min=4) and then another one that says AUTO(min=5)?

angelayi replied:
Does & know to take the min of 4 and 5? sorry.. im not familiar with &

pianpwk replied:
ah yep, the final thing would be the intersection of [2, inf] & [4, inf] & [5, inf], so we have a [5, inf] bound

tugsbayasgalan commented on torch/_export/non_strict_utils.py:
```diff
@@ -394,24 +399,53 @@ def make_constraints(
+                except torch.utils._sympy.value_ranges.ValueRangeError as exc:
+                    raise ValueError(
+                        f"Received user-specified min/max range of [{dim.min}, {dim.max}], "
+                        f"which conflicted with the inferred min/max range of [{trace_vr.lower}, {trace_vr.upper}] "
```
What happens when the inferred dim is static? I think this error message is going to be weird since the lower and upper are same numbers?

pianpwk replied:
Yeah, with AUTO this would report something like, user-specified [8, 12], while inferred [4, 4].

tugsbayasgalan replied:
I think we should suggest that it is static in this case, otherwise it is bit confusing

pianpwk committed:
[export] min/max ranges for dim hints (#149590)

Summary:
Adds min/max ranges to Dim.AUTO/DYNAMIC/STATIC, so users can do `Dim.AUTO(min=2, max=2048)`.


Reviewed By: tugsbayasgalan

Differential Revision: D71522032

Pulled By: pianpwk
